Compile/src/Producers/LeptonSFProducer.cc

```diff
@@ -468,3 +469,91 @@ void LeptonTriggerSFProducer::Produce(ZJetEvent const& event,
         }
     }
 }
+
+/////////////////////
+// LeptonRecoSF //
+/////////////////////
+
+std::string LeptonRecoSFProducer::GetProducerId() const { return "LeptonRecoSFProducer"; }
+
+void LeptonRecoSFProducer::Init(ZJetSettings const& settings)
+{
+    m_sffile = settings.GetLeptonRecoSFRootfile();
+    if (settings.GetLeptonSFVariation() == true) {
+        LOG(INFO) << "LeptonRecoSFProducer: varying scale factor UP and DOWN one sigma";
+    }
+
+    if(settings.GetChannel() == "mm"){
+        histoname = settings.GetLeptonRecoSFHistogramName();
+    }
+    else{
+        LOG(ERROR) << "LeptonRecoSFProducer not implemented for this channel";
+    }
+
+    GetEtaAxis2D(m_etarev2D,histoname);
+    m_absoluteeta   = m_etarev2D[0];
+    m_reversed_axes = m_etarev2D[1];
+    m_2D            = m_etarev2D[2];
+    if (m_reversed_axes)
+    {
+        m_etabins = &m_ybins;
+        m_ptbins = &m_xbins;
+    }
+    else
+    {
+        m_etabins = &m_xbins;
+        m_ptbins = &m_ybins;
+    }
+    // Get file
+    LOG(INFO) << "Loading lepton scale factors for Reco " << m_id << ": File " << m_sffile
+              << ", Histogram " << histoname;
+    TFile file(m_sffile.c_str(), "READONLY");
+    TH2F* sfhisto = (TH2F*)file.Get(histoname.c_str());
+
+    // Get the pT and eta bin borders
+    for (int ix = 0; ix <= sfhisto->GetNbinsX(); ++ix){
+        m_xbins.emplace_back(2 * sfhisto->GetXaxis()->GetBinCenter(ix) -
+                             sfhisto->GetXaxis()->GetBinLowEdge(ix));
+    }
+    for (int iy = 0; iy <= sfhisto->GetNbinsY(); ++iy){
+        m_ybins.emplace_back(2 * sfhisto->GetYaxis()->GetBinCenter(iy) -
+                             sfhisto->GetYaxis()->GetBinLowEdge(iy));
+    }
+    // Fill the m_sf array with the values from the root histo
+    for (int ix = 1; ix <= sfhisto->GetNbinsX(); ix++) {
+        for (int iy = 1; iy <= sfhisto->GetNbinsY(); iy++) {
+            m_sf[ix - 1][iy - 1] = static_cast<float>(sfhisto->GetBinContent(ix, iy));
+            m_er[ix - 1][iy - 1] = static_cast<float>(sfhisto->GetBinError(ix, iy));
+        }
+    }
+    file.Close();
+}
+void LeptonRecoSFProducer::Produce(ZJetEvent const& event,
+                                    ZJetProduct& product,
+                                    ZJetSettings const& settings) const
+{
+    if(product.m_zValid){
```

Comment:
I would suggest to directly add debug output here ;)
`LOG(DEBUG) << "\n[LeptonRecoSFProducer]\n";`

Reply:
Default is not to add the Producer. Reosonable defaults for root filename and histo name are not possible.

Reply:
I mean to add for example a warning, when the producer is executed but settings.GetLeptonSFVariation() is not set.

Reply:
The variation is already printed during initialization. But I'll add it here as well, so we can make sure it's working.

Reply:
:+1: